### ISSUE TITLE
Cannot use the index "mapBak1HD_sorted_name" of the IMap "mapBak1HD" due to concurrent migration, or because index creation is still in progress

### ISSUE THREAD

**Issue by Danny-Hazelcast** [hazelcast/hazelcast#18380]
No migration in progress.

dynamically add index to a field named name 

concurrent sql service queries running on fields,  "id" and "price"

with no migration in progress then adding an index on a field while running queries on unrelated fields gives exception 

"mapBak1HD_sorted_name" is not being used, the field "name" is not being used, while an index on name is dynamically created with no migration in progress

```
com.hazelcast.sql.HazelcastSqlException: Cannot use the index "mapBak1HD_sorted_name" of the IMap "mapBak1HD" due to concurrent migration, or because index creation is still in progress
	at com.hazelcast.sql.impl.client.SqlClientService.handleResponseError(SqlClientService.java:264)
	at com.hazelcast.sql.impl.client.SqlClientService.handleExecuteResponse(SqlClientService.java:133)
	at com.hazelcast.sql.impl.client.SqlClientService.lambda$execute$0(SqlClientService.java:111)
	at com.hazelcast.internal.util.ExceptionUtil.lambda$withTryCatch$1(ExceptionUtil.java:206)
	at com.hazelcast.client.impl.spi.impl.ClientInvocationFuture$CallIdTrackingBiConsumer.accept(ClientInvocationFuture.java:268)
	at com.hazelcast.spi.impl.AbstractInvocationFuture$WhenCompleteNode.lambda$execute$0(AbstractInvocationFuture.java:1563)
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

```
[danny@server16 sqlService]$ pwd
/home/danny/workspace/sqlService
```

**Comment by Danny-Hazelcast** [comment]
can you add an index to a field dynamically with out causing HazelcastSqlException to queries running on different fields
while no migration is in progress. ?

**Comment by Danny-Hazelcast** [comment]
-Dhazelcast.hd.global.index.enabled=true

**Comment by Danny-Hazelcast** [comment]
```
com.hazelcast.sql.HazelcastSqlException: Cannot resolve IMap schema because it doesn't have entries on the local member: mapBak1HD
	at com.hazelcast.sql.impl.client.SqlClientService.handleResponseError(SqlClientService.java:264)
	at com.hazelcast.sql.impl.client.SqlClientService.handleExecuteResponse(SqlClientService.java:133)
	at com.hazelcast.sql.impl.client.SqlClientService.lambda$execute$0(SqlClientService.java:111)
	at com.hazelcast.internal.util.ExceptionUtil.lambda$withTryCatch$1(ExceptionUtil.java:206)
	at com.hazelcast.client.impl.spi.impl.ClientInvocationFuture$CallIdTrackingBiConsumer.accept(ClientInvocationFuture.java:268)
	at com.hazelcast.spi.impl.AbstractInvocationFuture$WhenCompleteNode.lambda$execute$0(AbstractInvocationFuture.java:1563)
	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```


```
#!/usr/bin/env bash
hz broker $(hostname -i | awk '{print $5}')
hz jversion 8

ops="$ops -Dhazelcast.enterprise.license.key=${HAZELCAST_EE_KEY}"
#ops="$ops -Dhazelcast.partition.count=2411"
ops="$ops -Dhazelcast.partition.count=6007"
ops="$ops -Dhazelcast.hd.global.index.enabled=true"

hz xml config-hz-600
hz memberOps "-Xms12G -Xmx12G ${ops}"
hz clientOps "-Xms1G  -Xmx1G ${ops}"
hz cluster -ee -size M10C10 -v $(hazelcastVersion)

#sz=1000000000
sz=10000000

hz go hzcmd.map.loader.LoadEntitys threads=100 keyDomain=${sz} maxPrice=1000 name=mapBak1HD bench=No
hz go hzcmd.distributed.AddIndex driver=Member1 fields=id bench=No
#hz go hzcmd.distributed.AddIndex driver=Member1 fields=price bench=No
hz go hzcmd.distributed.Size driver=Member size=${sz} bench=No
hz go hzcmd.hd.Size driver=Member GB=100 bench=No


hz duration 0

#hz go -f A hzcmd.map.sql.predicate.IdRange keyDomain=${sz} range=1 name=mapBak1HD &
#hz go -f B hzcmd.map.sql.predicate.PriceIs keyDomain=${sz} maxPrice=1000 name=mapBak1HD &

hz go -f q-id hzcmd.map.sql.service.IdRange keyDomain=${sz} range=1 name=mapBak1HD &
#hz go -f q-price hzcmd.map.sql.service.PriceIs maxPrice=1000 name=mapBak1HD &
#hz go -f q-bool hzcmd.map.sql.service.BoolIs name=mapBak1HD &

sleep 30
hz go hzcmd.distributed.AddIndex driver=Member1 fields=name bench=No
hz go hzcmd.distributed.AddIndex driver=Member1 fields=price bench=No
hz go hzcmd.distributed.AddIndex driver=Member1 fields=uuid bench=No
hz go hzcmd.distributed.AddIndex driver=Member1 fields=strId bench=No
hz go hzcmd.distributed.AddIndex driver=Member1 fields=bool bench=No

wait
hz download
```


```
Installing core
installing /home/danny/.m2/repository/com/hazelcast/hazelcast-enterprise/4.2-SNAPSHOT/hazelcast-enterprise-4.2-SNAPSHOT.jar
installing /home/danny/.m2/repository/com/hazelcast/hazelcast-sql/4.2-SNAPSHOT/hazelcast-sql-4.2-SNAPSHOT.jar
HzMember1HZ Started 10.212.1.101 903295@Server1 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzMember4HZ Started 10.212.1.104 274156@SERVER4 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzMember8HZ Started 10.212.1.108 500962@SERVER8 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzMember7HZ Started 10.212.1.107 775224@SERVER7 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzMember10HZ Started 10.212.1.110 753539@server10 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzMember2HZ Started 10.212.1.102 272724@SERVER2 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzMember9HZ Started 10.212.1.109 499260@SERVER9 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzMember5HZ Started 10.212.1.105 614038@server5 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzMember6HZ Started 10.212.1.106 343465@server6 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzMember3HZ Started 10.212.1.103 903621@SERVER3 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzClient2HZ Started 10.212.1.102 272977@SERVER2 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzClient9HZ Started 10.212.1.109 499495@SERVER9 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzClient10HZ Started 10.212.1.110 753784@server10 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzClient3HZ Started 10.212.1.103 903904@SERVER3 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzClient5HZ Started 10.212.1.105 614273@server5 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzClient6HZ Started 10.212.1.106 343717@server6 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzClient1HZ Started 10.212.1.101 903560@Server1 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzClient7HZ Started 10.212.1.107 775478@SERVER7 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzClient8HZ Started 10.212.1.108 501200@SERVER8 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
HzClient4HZ Started 10.212.1.104 274393@SERVER4 /appdisk/danny/hz-root/hz-lib/4.2-SNAPSHOT
fn hzcmd.map.loader.LoadEntitys driver=.* threads=100 duration=0 bench=No interval=0 keyDomain=10000000 maxPrice=1000 name=mapBak1HD
fn Done
fn hzcmd.distributed.AddIndex driver=Member1 threads=1 duration=0 bench=No interval=0 fields=id
fn Done
fn hzcmd.distributed.Size driver=Member threads=1 duration=0 bench=No interval=0 size=10000000
fn Done
fn hzcmd.hd.Size driver=Member threads=1 duration=0 bench=No interval=0 GB=100
fail HzMember2HZ fn hzcmd.hd.Size threadId=0 global.AssertionException: used 1.29 < 100.0 GB's
q-id hzcmd.map.sql.service.IdRange driver=.* threads=1 duration=0 bench=Roll interval=0 keyDomain=10000000 range=1 name=mapBak1HD
fn hzcmd.distributed.AddIndex driver=Member1 threads=1 duration=0 bench=No interval=0 fields=name
fail HzClient7HZ q-id hzcmd.map.sql.service.IdRange threadId=0 com.hazelcast.sql.HazelcastSqlException: Cannot resolve IMap schema because it doesn't have entries on the local member: mapBak1HD
fn Done
fn hzcmd.distributed.AddIndex driver=Member1 threads=1 duration=0 bench=No interval=0 fields=price
fn Done
fn hzcmd.distributed.AddIndex driver=Member1 threads=1 duration=0 bench=No interval=0 fields=uuid
fn Done
fn hzcmd.distributed.AddIndex driver=Member1 threads=1 duration=0 bench=No interval=0 fields=strId
fn Done
fn hzcmd.distributed.AddIndex driver=Member1 threads=1 duration=0 bench=No interval=0 fields=bool
fn Done
```


queries on "id" field 
client get com.hazelcast.sql.HazelcastSqlException after index add on "name"

**Comment by Danny-Hazelcast** [comment]
```
#!/usr/bin/env bash
#set -e
echo $$ > pid.txt

hz broker $(hostname -i | awk '{print $5}')
hz jversion 8
#hz allFail

ops="$ops -Dhazelcast.enterprise.license.key=${HAZELCAST_EE_KEY}"
ops="$ops -Dhazelcast.partition.count=6007"
ops="$ops -Dhazelcast.hd.global.index.enabled=true"

hz xml config-hz-600
hz memberOps "-Xms12G -Xmx12G ${ops}"
hz clientOps "-Xms1G  -Xmx1G ${ops}"
hz cluster -ee -size M10C10 -v $(hazelcastVersion)
sz=10000000
for i in {1..10}; do
	hz go hzcmd.map.loader.LoadEntitys threads=100 keyDomain=${sz} maxPrice=1000 name=mapBak1HD bench=No
	hz go hzcmd.distributed.Size driver=Member size=${sz} bench=No
	hz go hzcmd.distributed.AddIndex driver=Member1 fields=id bench=No
	hz go -f q-id hzcmd.map.sql.service.IdRange keyDomain=${sz} range=1 name=mapBak1HD &
	sleep 4
	hz go hzcmd.distributed.AddIndex driver=Member1 fields=name bench=No
	hz go hzcmd.distributed.AddIndex driver=Member1 fields=price bench=No
	hz go hzcmd.distributed.AddIndex driver=Member1 fields=uuid bench=No
	hz go hzcmd.distributed.AddIndex driver=Member1 fields=strId bench=No
	hz go hzcmd.distributed.AddIndex driver=Member1 fields=bool bench=No
	hz stop
        wait
	hz go hzcmd.distributed.DestroyMaps driver=Member1 bench=No
done
```

**Comment by viliam-durina** [comment]
The issue was fixed by #18389 only partially: it fixes if a new index is added while there already was an existing index. It will fail the same if there was no index on the map before.

**Comment by viliam-durina** [comment]
This issue applied to IMDG engine which will be retired in 5.0
